src/db/db.c

```diff
@@ -452,6 +461,113 @@ dbBackupStop(Db *this)
     FUNCTION_LOG_RETURN_STRUCT(result);
 }
 
+/**********************************************************************************************************************************/
+#define DB_SYNC_TOTAL                                               16
+#define DB_SYNC_SIZE                                                32
+#define DB_SYNC_DATA                                                "elIlg6P7i1yQYtOfsfA4tNSn2io6QK18MzJwzjkBL2QpjuI"
+
+void
+dbSyncCheckHelper(PgClient *const pgClient, const String *const path)
+{
+    FUNCTION_LOG_BEGIN(logLevelDebug);
+        FUNCTION_LOG_PARAM(PG_CLIENT, pgClient);
+        FUNCTION_LOG_PARAM(STRING, path);
+    FUNCTION_LOG_END();
+
+    ASSERT(pgClient != NULL);
+    ASSERT(path != NULL);
+    ASSERT(sizeof(DB_SYNC_DATA) == DB_SYNC_TOTAL + DB_SYNC_SIZE);
+
+    MEM_CONTEXT_TEMP_BEGIN()
+    {
+        // Build data
+        const char *const file = strZ(strNewFmt("%s/pgbackrest.sync.tmp", strZ(path)));
+        const char *fileExpected[DB_SYNC_TOTAL];
+        const String *fileCmd[DB_SYNC_TOTAL];
+
+        for (unsigned int fileIdx = 0; fileIdx < DB_SYNC_TOTAL; fileIdx++)
+        {
+            fileExpected[fileIdx] = strZ(strNewN(DB_SYNC_DATA + fileIdx, DB_SYNC_SIZE));
+            fileCmd[fileIdx] = strNewFmt("copy (select '%s') to '%s'", fileExpected[fileIdx], file);
+
+            ASSERT(fileIdx == 0 || strncmp(fileExpected[fileIdx], fileExpected[fileIdx - 1], DB_SYNC_SIZE) != 0);
+        }
+
+        // Write data and collect results
+        struct stat fileStat[DB_SYNC_TOTAL];
+        char fileActual[DB_SYNC_TOTAL][DB_SYNC_SIZE];
+        ssize_t fileActualSize[DB_SYNC_TOTAL];
+
+        for (unsigned int fileIdx = 0; fileIdx < DB_SYNC_TOTAL; fileIdx++)
+        {
+            // Write data into the file using a copy on the pg server
+            CHECK(pgClientQuery(pgClient, fileCmd[fileIdx]) == NULL);
+
+            // Stat the file
+            THROW_ON_SYS_ERROR_FMT(stat(file, &fileStat[fileIdx]) == -1, FileReadError, "unable to stat '%s'", file);
+
+            // Read the file
+            int fd = open(file, O_RDONLY, 0);
+            THROW_ON_SYS_ERROR_FMT(fd == -1, FileOpenError, "unable to open '%s'", file);
+
+            fileActualSize[fileIdx] = read(fd, fileActual[fileIdx], DB_SYNC_SIZE);
+            THROW_ON_SYS_ERROR_FMT(fileActualSize[fileIdx] == -1, FileReadError, "unable to read '%s'", file);
+
+            THROW_ON_SYS_ERROR_FMT(close(fd) == -1, FileCloseError, "unable to close '%s'", file);
+        }
+
+        // Check results
+        for (unsigned int fileIdx = 0; fileIdx < DB_SYNC_TOTAL; fileIdx++)
+        {
+            CHECK(fileStat[fileIdx].st_size == DB_SYNC_SIZE + 1);
```

Comment:
This assumes that the end of line is exactly one byte and it doesn't check that it is that EOL character instead of some other junk. I think a better way would be to make this read buffer larger by 4 bytes, attempt the read in that size, compare that the number of bytes actually read matches st_size from fstat(2), then right-strip every '\r' and '\n' character and strcmp(3) the result.

The reason for allocating 4 more bytes is that a read(2) does not NUL terminate the string, so we need to first make sure that read(2) in the length of DB_SYNC_SIZE+4 did return less bytes than that and NUL terminate the data at the offset, read(2) returned. We expect an end-of-line sequence, which may be LF or CRLF, but handle the case that there could be some garbage after that.

We may not expect to run on anything but Unix and I am not aware of any filesystem or setting that could cause the end-of-line to be CRLF. But that could change tomorrow.